File: ts-playwright/src/api/BookerClient.ts
import { APIRequestContext, APIResponse } from '@playwright/test';

export interface Booking {
  firstname: string;
  lastname: string;
  totalprice: number;
  depositpaid: boolean;
  bookingdates: { checkin: string; checkout: string };
  additionalneeds?: string;
}

/**
 * Thin client for the restful-booker API.
 * Quirks: POST /auth returns a token (sent as a cookie for update/delete);
 * DELETE returns 201 on success.
 */
export class BookerClient {
  private request: APIRequestContext;
  private token: string | null = null;

  constructor(request: APIRequestContext) {
    this.request = request;
  }

  async auth(username: string, password: string): Promise<string> {
    const resp = await this.request.post('/auth', { data: { username, password } });
    if (!resp.ok()) throw new Error(`auth failed: ${resp.status()} ${await resp.text()}`);
    this.token = (await resp.json()).token;
    return this.token!;
  }

  private authHeaders(): Record<string, string> {
    if (!this.token) throw new Error('Call auth() before update/delete.');
    return {
      Cookie: `token=${this.token}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };
  }

  createBooking(booking: Booking): Promise<APIResponse> {
    return this.request.post('/booking', {
      data: booking,
      headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    });
  }

  getBooking(id: number): Promise<APIResponse> {
    return this.request.get(`/booking/${id}`, { headers: { Accept: 'application/json' } });
  }

  updateBooking(id: number, booking: Booking): Promise<APIResponse> {
    return this.request.put(`/booking/${id}`, { data: booking, headers: this.authHeaders() });
  }

  deleteBooking(id: number): Promise<APIResponse> {
    return this.request.delete(`/booking/${id}`, { headers: this.authHeaders() });
  }
}

File: python-playwright/framework/api/booker_client.py
"""Thin client for the restful-booker API, built on Playwright's
``APIRequestContext``.

restful-booker quirks worth knowing:
- ``POST /auth`` returns a ``token`` you must send as a cookie to update/delete.
- Write endpoints want ``Content-Type: application/json`` and
  ``Accept: application/json``.
- ``DELETE`` returns ``201 Created`` (not 200/204) on success.
"""

from __future__ import annotations

from typing import Any

from playwright.sync_api import APIRequestContext


class BookerClient:
    def __init__(self, request: APIRequestContext) -> None:
        self.request = request
        self._token: str | None = None

    # ---- auth ---- #
    def auth(self, username: str, password: str) -> str:
        resp = self.request.post("/auth", data={"username": username, "password": password})
        assert resp.ok, f"auth failed: {resp.status} {resp.text()}"
        token: str = resp.json()["token"]
        self._token = token
        return token

    @property
    def _auth_headers(self) -> dict[str, str]:
        if not self._token:
            raise RuntimeError("Call auth() before update/delete.")
        return {
            "Cookie": f"token={self._token}",
            "Content-Type": "application/json",
            "Accept": "application/json",
        }

    # ---- bookings ---- #
    def create_booking(self, booking: dict[str, Any]):
        return self.request.post(
            "/booking",
            data=booking,
            headers={"Content-Type": "application/json", "Accept": "application/json"},
        )

    def get_booking(self, booking_id: int):
        return self.request.get(f"/booking/{booking_id}", headers={"Accept": "application/json"})

    def update_booking(self, booking_id: int, booking: dict[str, Any]):
        return self.request.put(f"/booking/{booking_id}", data=booking, headers=self._auth_headers)

    def delete_booking(self, booking_id: int):
        return self.request.delete(f"/booking/{booking_id}", headers=self._auth_headers)
